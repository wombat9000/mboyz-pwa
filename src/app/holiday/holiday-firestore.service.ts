import {Injectable} from '@angular/core';
import {User} from '../core/auth.service';
import {Observable} from 'rxjs/Observable';
import {AngularFirestore, AngularFirestoreCollection} from 'angularfire2/firestore';
import {Holiday} from './holiday.service';


interface FSHoliday {
  id: string;
  name: string;
}


@Injectable()
export class HolidayFirestore {


  holidaysCollection: AngularFirestoreCollection<Holiday>;
  holidays: Observable<Holiday[]>;

  constructor(private afs: AngularFirestore) {
    this.holidaysCollection = this.afs.collection('holidays');
    this.holidays = this.holidaysCollection.valueChanges();
  }

  public observeById(holidayId: string): Observable<Holiday> {
    return this.afs.doc<Holiday>(`holidays/${holidayId}`).valueChanges();
  }

  public save(holiday: Holiday) {
    const data = {
      id: holiday.id,
      name: holiday.name
    };

    return this.afs.doc(`holidays/${holiday.id}`).set(data);
  }
}
