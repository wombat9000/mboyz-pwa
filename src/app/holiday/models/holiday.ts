import {now} from 'moment';
import {DbRecord} from './DbRecord';

export interface HolidayDTO extends DbRecord {
  readonly name: string;
  readonly created: string;
}

export function newTestHoliday(id: string): HolidayDTO {
  return {
    id: id,
    name: 'someName',
    created: now().toString()
  };
}
