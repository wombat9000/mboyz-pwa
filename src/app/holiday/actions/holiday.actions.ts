import {Action} from '@ngrx/store';
import {Holiday} from '../holiday.service';


export const QUERY    = '[Holiday] query pizzas';
export const ADDED    = '[Holiday] added';
export const MODIFIED = '[Holiday] modified';
export const REMOVED  = '[Holiday] removed';
export const CREATE  = '[Holiday] created';

// Initial query
export class Query implements Action {
  readonly type = QUERY;
  constructor() {}
}
// AngularFire2 StateChanges
export class Added implements Action {
  readonly type = ADDED;
  constructor(public holiday: Holiday) {}
}

export class Create implements Action {
  readonly type = CREATE;
  constructor(public holiday: Holiday) {}
}

export class Modified implements Action {
  readonly type = MODIFIED;
  constructor(public holiday: Holiday) {}
}

export class Removed implements Action {
  readonly type = REMOVED;
  constructor(public holiday: Holiday) {}
}


export type HolidayActions =
  Create |
  Query |
  Added |
  Modified |
  Removed;
