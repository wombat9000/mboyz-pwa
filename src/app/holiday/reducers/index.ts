import {createFeatureSelector, createSelector} from '@ngrx/store';
import * as fromHolidays from './holiday.reducer';
import * as fromRoot from '../../reducers';


export interface HolidaysState {
  holidays: fromHolidays.State;
}

export interface State extends fromRoot.State {
  holidays: HolidaysState;
}

export const reducers = {
  holidays: fromHolidays.reducer
};

export const getHolidaysState = createFeatureSelector<HolidaysState>('holidays');

export const getHolidayEntitiesState = createSelector(
  getHolidaysState,
  state => state.holidays
);

export const getSelectedHolidayId = createSelector(
  getHolidayEntitiesState,
  fromHolidays.getSelectedId
);

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = fromHolidays.adapter.getSelectors(getHolidayEntitiesState);

export const getSelectedHoliday = createSelector(
  selectEntities,
  getSelectedHolidayId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);
