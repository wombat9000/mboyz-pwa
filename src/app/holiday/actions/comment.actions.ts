import {Action} from '@ngrx/store';
import {MbComment} from '../models/comment';

export const CREATE         = '[comments CommentBox] create';
export const AF_ADDED       = '[comments Firestore] added';
export const AF_MODIFIED    = '[comments Firestore] modified';
export const AF_REMOVED     = '[comments Firestore] removed';


export class Create implements Action {
  readonly type = CREATE;

  constructor(public payload: { comment: MbComment }) {
  }
}

export class AfAdded implements Action {
  readonly type = AF_ADDED;

  constructor(readonly payload: { comment: MbComment }) {
  }
}

export class AfModified implements Action {
  readonly type = AF_MODIFIED;

  constructor(readonly payload: { comment: Partial<MbComment> }) {
  }
}

export class AfRemoved implements Action {
  readonly type = AF_REMOVED;

  constructor(readonly payload: { comment: MbComment }) {
  }
}

export type CommentActions =
  Create |
  AfAdded |
  AfModified |
  AfRemoved;

