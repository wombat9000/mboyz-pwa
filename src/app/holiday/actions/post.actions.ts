import {Action} from '@ngrx/store';
import {PostDTO} from '../models/post';
import {CreateAction} from '../../core/actions/data.actions';

export const CREATE = '[posts PostboxComponent] create';
export const CREATE_SUCCESS = '[posts App] create success';
export const AF_ADDED = '[posts Firestore] added';
export const AF_MODIFIED = '[posts Firestore] modified';
export const AF_REMOVED = '[posts Firestore] removed';

export class Create implements CreateAction {
  readonly type = CREATE;

  constructor(public readonly payload: { record: PostDTO }) {
  }
}

export class CreateSuccess implements Action {
  readonly type = CREATE_SUCCESS;

  constructor() {
  }
}

export class AfAdded implements CreateAction {
  readonly type = AF_ADDED;

  constructor(public readonly payload: { record: PostDTO }) {
  }
}

export class AfModified implements Action {
  readonly type = AF_MODIFIED;

  constructor(public readonly payload: { record: Partial<PostDTO> }) {
  }
}

export class AfRemoved implements Action {
  readonly type = AF_REMOVED;

  constructor(public readonly payload: { record: PostDTO }) {
  }
}

export type PostActions =
  Create |
  AfAdded |
  AfModified |
  AfRemoved;

