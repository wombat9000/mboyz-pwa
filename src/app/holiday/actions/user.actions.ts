import {Action} from '@ngrx/store';
import {CreateAction} from '../../core/actions/data.actions';
import {MtravelUser} from '../../auth/services/auth.service';

export const CREATE = '[users PostboxComponent] create';
export const CREATE_SUCCESS = '[users App] create success';
export const AF_ADDED = '[users Firestore] added';
export const AF_MODIFIED = '[users Firestore] modified';
export const AF_REMOVED = '[users Firestore] removed';

export class Create implements CreateAction {
  readonly type = CREATE;

  constructor(public readonly payload: { record: MtravelUser }) {
  }
}

export class CreateSuccess implements Action {
  readonly type = CREATE_SUCCESS;

  constructor() {
  }
}

export class AfAdded implements CreateAction {
  readonly type = AF_ADDED;

  constructor(public readonly payload: { record: MtravelUser }) {
  }
}

export class AfModified implements Action {
  readonly type = AF_MODIFIED;

  constructor(public readonly payload: { record: Partial<MtravelUser> }) {
  }
}

export class AfRemoved implements Action {
  readonly type = AF_REMOVED;

  constructor(public readonly payload: { record: MtravelUser }) {
  }
}

export type MtravelUserActions =
  Create |
  AfAdded |
  AfModified |
  AfRemoved;

