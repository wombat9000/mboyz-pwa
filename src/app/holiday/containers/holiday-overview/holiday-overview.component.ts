import {Component, OnInit} from '@angular/core';
import {Router} from '@angular/router';
import {Observable} from 'rxjs';
import {Store} from '@ngrx/store';
import * as fromHoliday from '../../reducers';
import {Holiday} from '../../models/holiday';
import * as moment from 'moment';
import {map} from 'rxjs/operators';
import {SetTitle} from '../../../core/actions/app-bar.actions';

@Component({
  selector: 'app-holiday-overview',
  template: `
    <button
      class="add-holiday"
      mat-fab
      routerLink="/holiday/create">
      <mat-icon>add</mat-icon>
    </button>


    <mat-list role="list">
  <span *ngFor="let holiday of holidays$ | async">
    <mat-list-item role="listitem" (click)="goToDetail(holiday.id)">
      <div class="holiday-item">
      <div>
        <span class="holiday-name">{{holiday.name}}</span>
      </div>

      <div>
        <span class="created-text">Erzeugt am {{formatDate(holiday.created)}} um {{formatTime(holiday.created)}}</span>
      </div>
        </div>
    </mat-list-item>
    <mat-divider></mat-divider>
  </span>
    </mat-list>
  `,
  styleUrls: ['./holiday-overview.component.scss']
})
export class HolidayOverviewPageComponent implements OnInit {

  holidays$: Observable<Holiday[]>;

  constructor(private store: Store<fromHoliday.State>,
              private router: Router) {
  }

  ngOnInit() {
    this.holidays$ = this.store.select(fromHoliday.selectAll).pipe(
      map(this.sortByDate)
    );

    // TODO:test
    this.store.dispatch(new SetTitle({newTitle: 'holidays'}));
  }

  goToDetail(holidayId: string) {
    return this.router.navigateByUrl(`/holiday/${holidayId}`);
  }

  formatDate(isoString: string): string {
    return moment(isoString).format('Do MMMM');
  }

  formatTime(isoString: string): string {
    return moment(isoString).format('LT');
  }

  private sortByDate(holidays: Holiday[]) {
    return holidays
      .filter(it => it.created !== undefined)
      .sort((some, other) => {
        return moment(some.created).isAfter(moment(other.created)) ? 0 : 1;
      });
  }
}
