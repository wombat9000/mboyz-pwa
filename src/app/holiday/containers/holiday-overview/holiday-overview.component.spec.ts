import {async, ComponentFixture, TestBed} from '@angular/core/testing';

import {HolidayOverviewPageComponent} from './holiday-overview.component';
import {By} from '@angular/platform-browser';
import {CUSTOM_ELEMENTS_SCHEMA, DebugElement} from '@angular/core';
import {Router} from '@angular/router';
import {routerMocker} from '../../../test-support/stubs';
import {click} from '../../../test-support/functions';
import {combineReducers, Store, StoreModule} from '@ngrx/store';

import * as fromHoliday from '../../reducers';
import * as actions from '../../actions/holiday.actions';
import * as moment from 'moment';

describe('HolidayOverviewComponent', () => {
  let component: HolidayOverviewPageComponent;
  let debugElement: DebugElement;
  let fixture: ComponentFixture<HolidayOverviewPageComponent>;

  let router;
  let store: Store<fromHoliday.State>;

  const firstHoliday = {id: 'someId', name: 'first created holiday', created: moment('2018-12-01').toISOString()};
  const moreRecentHoliday = {
    id: 'anotherId',
    name: 'more recently created holiday',
    created: moment('2018-12-02').toISOString()
  };


  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          holidays: combineReducers(fromHoliday.reducers)
        })
      ],
      providers: [
        {provide: Router, useFactory: routerMocker},
      ],
      declarations: [HolidayOverviewPageComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    }).compileComponents();
  }));

  beforeEach(() => {
    router = TestBed.get(Router);
    store = TestBed.get(Store);

    store.dispatch(new actions.Create(firstHoliday));
    store.dispatch(new actions.Create(moreRecentHoliday));

    spyOn(store, 'dispatch');

    fixture = TestBed.createComponent(HolidayOverviewPageComponent);
    component = fixture.componentInstance;
    debugElement = fixture.debugElement;
    fixture.detectChanges();
  });

  describe('displays holidays', () => {
    beforeEach(async () => {

      await fixture.whenStable();
      fixture.detectChanges();
    });

    it('should query for holiday changes', () => {
      expect(store.dispatch).toHaveBeenCalledWith(new actions.Query());
    });

    it('should display all holidays, sorted by date', () => {
      const cards = debugElement.queryAll(By.css('.holiday-name'))
        .map(it => it.nativeElement.textContent);

      expect(cards).toEqual([moreRecentHoliday.name, firstHoliday.name]);
    });

    describe('onClick', () => {
      it('should take user to holiday detail', () => {
        const someHolidayCard = debugElement.queryAll(By.css('.holiday-name'))
          .find(it => it.nativeElement.textContent === firstHoliday.name);

        click(someHolidayCard.nativeElement);

        const destinationURL = router.navigateByUrl.calls.first().args[0];
        expect(destinationURL).toBe(`/holiday/${firstHoliday.id}`);
      });
    });
  });

  describe('add button', () => {
    it('should show an add button', () => {
      const addButton = debugElement.query(By.css('.add-holiday'));
      expect(addButton).toBeTruthy();
    });

    it('should redirect to create holiday page onclick', () => {
      const addButton = debugElement.query(By.css('.add-holiday'));

      click(addButton);

      const destinationURL = router.navigateByUrl.calls.first().args[0];
      expect(destinationURL).toBe('/holiday/create');
    });
  });
});
