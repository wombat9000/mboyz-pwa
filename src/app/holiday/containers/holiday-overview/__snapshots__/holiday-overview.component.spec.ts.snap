// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HolidayOverviewComponent displays holidays should compile 1`] = `
<app-holiday-overview
  holidays$={[Function Store]}
  router={[Function Object]}
  store={[Function Store]}
>
  <button
    class="add-holiday"
    mat-fab=""
    routerlink="/holiday/create"
  >
    <mat-icon>
      add
    </mat-icon>
  </button><mat-list
    role="list"
  >
    
    <span>
      <mat-list-item
        role="listitem"
      >
        <div
          class="holiday-item"
        >
          <div>
            <span
              class="holiday-name"
            >
              more recently created holiday
            </span>
          </div>
          <div>
            <span
              class="created-text"
            >
              Erzeugt am 2nd December um 12:00 AM
            </span>
          </div>
        </div>
      </mat-list-item>
      <mat-divider />
    </span>
    <span>
      <mat-list-item
        role="listitem"
      >
        <div
          class="holiday-item"
        >
          <div>
            <span
              class="holiday-name"
            >
              first created holiday
            </span>
          </div>
          <div>
            <span
              class="created-text"
            >
              Erzeugt am 1st December um 12:00 AM
            </span>
          </div>
        </div>
      </mat-list-item>
      <mat-divider />
    </span>
  </mat-list>
</app-holiday-overview>
`;
